### PR TITLE
Fix slack report failing for doctest

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

By truncating the long reports.

The failing of the (a bit) many TF models is not related to `torch 2.2`. Previous we didn't see those failures, as `Blip2` causes OOM, and later tests may get more GPU memory due to this. (This is what I observed by re-run the doctest with different commits/torc versions).

Still need to figure if there is a way to avoid this.

### CODE REVIEW

File: utils/notification_service_doc_tests.py [hunk 1]
```diff
@@ -133,21 +133,38 @@ def failures(self) -> Dict:
...
+            return text
+
+        report = ""
+        for category, failures in category_failures.items():
+            max_len = MAX_ERROR_TEXT - len(report) - len("\n[Truncated]")
+            new_text = single_category_failures(category, failures, target_max_len=max_len, report=report)
+            if len(new_text) > max_len:
+                report += "\n[Truncated]"
+                break
+            report += new_text
```
Comment (ydshieh) [inline utils/notification_service_doc_tests.py:167]: A bit more complex than before: here we have 2 for loops: outer loop for categories, and inner loop for each single failure in a specific category.

**Comment by LysandreJik** [comment]
Maybe we can send two reports instead of truncating? WDYT?

**Comment by LysandreJik** [comment]
I meant that after truncating the first report, you could take the truncated text and send it over in a second report as a follow up. This way we would have all the information held in the reports

**Comment by ydshieh** [comment]
I made the `API Examples failures` and `MD Examples failures:` as independent slack `blocks`. Each report could be truncated.

See the following screenshot: (here there is no truncation, as I just ran 4 tests to see the output)

Screenshot 2023-10-27 105256

(I didn't try to customize the `The following examples had failures` part for each category however - to keep it simple)

File: utils/notification_service.py [hunk 2]
```diff
@@ -314,7 +314,7 @@ def compute_diff_for_failure_reports(self, curr_failure_report, prev_failure_rep
         return entries_changed
 
     @property
-    def model_failures(self) -> Dict:
+    def model_failures(self) -> List[Dict]:
```
Comment (ydshieh) [inline utils/notification_service.py:317]: Just fix the return type (forgot to update when I made change this block before)

File: utils/notification_service_doc_tests.py [hunk 3]
```diff
@@ -132,47 +132,43 @@ def failures(self) -> Dict:
...
+            if len(report) == 0:
+                continue
+            block = {
+                "type": "section",
+                "text": {
+                    "type": "mrkdwn",
+                    "text": f"The following examples had failures:\n\n\n{report}\n",
+                },
+            }
+            failure_blocks.append(block)
```
Comment (ydshieh) [inline utils/notification_service_doc_tests.py:169]: each category has its own block (for slack API `client.chat_postMessage`)

**Review by LysandreJik — APPROVED** [review]
Great, thanks @ydshieh!

**Pull request merged**